client/service/src/chain_ops.rs

```diff
@@ -392,18 +394,24 @@ impl<
 									}
 								}
 								Err(e) => {
-									return std::task::Poll::Ready(
+									return Poll::Ready(
 										Err(Error::Other(format!("Error reading block #{}: {}", read_block_count, e))))
 								}
 							}
 						}
 					}
 				},
-				ImportState::WaitingForImportQueueToCatchUp{block_iter, delay, block} => {
+				ImportState::WaitingForImportQueueToCatchUp{block_iter, mut delay, block} => {
 					let read_block_count = block_iter.read_block_count();
 					if read_block_count - link.imported_blocks >= MAX_PENDING_BLOCKS {
-						thread::sleep(delay);
 						// Queue is still full, so wait until there is room to insert our block.
+						match Pin::new(&mut delay).poll(cx) {
+							Poll::Pending => {
+								state = Some(ImportState::WaitingForImportQueueToCatchUp{block_iter, delay, block});
+								return Poll::Pending
+							},
+							Poll::Ready(_) => {},
```

Comment:
You need to create a new timer, otherwise you run into a hot loop here.

Reply:
(Delay, to be more specific)

Reply:
Do I need to create a new one, or would [using the reset method](https://docs.rs/futures-timer/3.0.2/futures_timer/struct.Delay.html#method.reset) work too?